File: android/src/main/java/com/thirddigital/exceptiontracking/ExceptionTrackingPluginPlugin.java
package com.thirddigital.exceptiontracking;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "NativeExceptionHandler")
public class ExceptionTrackingPluginPlugin extends Plugin {

    private final ExceptionTrackingPlugin implementation = new ExceptionTrackingPlugin();

    @Override
    public void load() {
        implementation.attach(getContext(), this);
    }

    @PluginMethod
    public void configure(PluginCall call) {
        try {
            implementation.configure(getContext(), this, call.getData());
            call.resolve();
        } catch (Exception exception) {
            call.reject("Failed to configure native exception handler", exception);
        }
    }

    @PluginMethod
    public void releaseExceptionHold(PluginCall call) {
        boolean handled = call.getBoolean("handled", true);
        implementation.releaseExceptionHold(handled);
        call.resolve();
    }

    @PluginMethod
    public void uploadPendingException(PluginCall call) {
        boolean uploaded = implementation.uploadPendingException(getContext());
        JSObject ret = new JSObject();
        ret.put("uploaded", uploaded);
        call.resolve(ret);
    }

    @PluginMethod
    public void crashForTesting(PluginCall call) {
        String message = call.getString("message", "Test native exception from Capacitor");
        implementation.crashForTesting(message);
        call.resolve();
    }

    void emitNativeException(JSObject event) {
        notifyListeners("nativeException", event, true);
    }
}

File: ios/Sources/ExceptionTrackingPluginPlugin/ExceptionTrackingPluginPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(ExceptionTrackingPluginPlugin)
public class ExceptionTrackingPluginPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "NativeExceptionHandler"
    public let jsName = "NativeExceptionHandler"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "configure", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setContext", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "releaseExceptionHold", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "uploadPendingException", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "crashForTesting", returnType: CAPPluginReturnPromise)
    ]
    private let implementation = ExceptionTrackingPlugin.shared

    public override func load() {
        implementation.attach(plugin: self)
    }

    @objc func configure(_ call: CAPPluginCall) {
        implementation.configure(call)
        call.resolve()
    }

    @objc func setContext(_ call: CAPPluginCall) {
        implementation.setContext(call)
        call.resolve()
    }

    @objc func releaseExceptionHold(_ call: CAPPluginCall) {
        implementation.releaseExceptionHold(handled: call.getBool("handled") ?? true)
        call.resolve()
    }

    @objc func uploadPendingException(_ call: CAPPluginCall) {
        call.resolve([
            "uploaded": implementation.uploadPendingException()
        ])
    }

    @objc func crashForTesting(_ call: CAPPluginCall) {
        implementation.crashForTesting(message: call.getString("message") ?? "Test native exception from Capacitor")
        call.resolve()
    }

    func emitNativeException(_ event: [String: Any]) {
        notifyListeners("nativeException", data: event, retainUntilConsumed: true)
    }
}
